File: fullstackProject/apps/api/src/rabbit/retry.ts
import type { MessageErrorHandler } from '@golevelup/nestjs-rabbitmq';
import { Logger, NotFoundException } from '@nestjs/common';
import type { Channel, ConsumeMessage } from 'amqplib';

/** заголовок-счётчик попыток: инкрементим при каждом уходе в retry */
export const RETRY_HEADER = 'x-retry-count';

/** всего попыток обработки (1 оригинал + 2 ретрая), дальше — parking */
export const MAX_ATTEMPTS = 3;

/** сколько retry-очередь держит сообщение перед возвратом в рабочую */
export const RETRY_TTL_MS = 5_000;

/** сколько раз сообщение уже ретраилось (0 — оригинал) */
export function readAttempts(msg: ConsumeMessage): number {
  const value = msg.properties.headers?.[RETRY_HEADER];
  return typeof value === 'number' ? value : 0;
}

/**
 * ErrorHandler для @RabbitSubscribe: сообщение при сбое обработки не теряется.
 *
 * Транзиентная ошибка (например, БД недоступна) → публикация копии в
 * `<rk>.retry`: эта очередь держит сообщение RETRY_TTL_MS и по TTL сама
 * возвращает его в рабочую очередь — паузу между попытками делает брокер.
 * «Ядовитое» сообщение (NotFound — такой брони уже нет) ретраить бессмысленно,
 * оно уходит в parking сразу. После MAX_ATTEMPTS попыток в `<rk>.parking`
 * уходит всё — эти сообщения разбирают вручную (админ-очередь).
 */
export function retryErrorHandler(routingKey: string): MessageErrorHandler {
  const logger = new Logger(`RabbitRetry[${routingKey}]`);
  return (channel: Channel, msg: ConsumeMessage, error: unknown) => {
    const attempt = readAttempts(msg) + 1;
    const poison = error instanceof NotFoundException;
    const target = !poison && attempt < MAX_ATTEMPTS ? 'retry' : 'parking';
    const message = error instanceof Error ? error.message : String(error);
    logger.warn(`попытка ${attempt}: ${message} → ${target}`);

    channel.publish('cinema', `${routingKey}.${target}`, msg.content, {
      contentType: msg.properties.contentType,
      persistent: true,
      headers: {
        ...msg.properties.headers,
        [RETRY_HEADER]: attempt,
        'x-last-error': message,
      },
    });
    channel.ack(msg); // судьба исходного решена: оно уехало в retry/parking
  };
}

File: fullstackProject/services/ticket-worker/internal/rabbitmq/retry.go
package rabbitmq

import (
	amqp "github.com/rabbitmq/amqp091-go"
)

// retryHeader — счётчик попыток; растёт с каждым уходом в retry-очередь.
const retryHeader = "x-retry-count"

// attempts читает счётчик попыток из заголовков доставки (0 — оригинал).
// AMQP-типы целого зависят от брокера, поэтому принимаем всё int-семейство.
func attempts(d amqp.Delivery) int {
	switch v := d.Headers[retryHeader].(type) {
	case int:
		return v
	case int16:
		return int(v)
	case int32:
		return int(v)
	case int64:
		return int(v)
	default:
		return 0
	}
}

// routeFor решает судьбу упавшего сообщения: транзиентная ошибка ретраится,
// пока попыток меньше max; «ядовитое» (poison) едет в parking сразу —
// ретраи его не исправят.
func routeFor(attempt, max int, poison bool) string {
	if poison || attempt >= max {
		return "parking"
	}
	return "retry"
}
